### DISCUSSION TITLE
Separate Queueable runs are collected into one Log__c record

### DISCUSSION THREAD

**Discussion by reitffunk** [jongpie/NebulaLogger#310] [Q&A] [answered]
Hi,

I have a `Queueable` where I also set the `parentLogTransactionId` accordingly to the provided examples. At the end of this `Queueable` the same is chained again until there is no work left.

Regarding to the logs I would have expected that every separate execution of the `Queueable` creates its own `Log__c` record.

Instead all the runs are collected into one `Log__c` record. You can see it in the screenshot where on the right side it says for the first bunch of entries `1` and for the second bunch `2`. These are basically the separate executions of the `Queueable` and every `1`, `2` should be a separate `Log__c` record.

The funny thing: Somehow the same code produces multiple Log__c records within the Sandbox.

I do not understand how they are related at all as every separate `Queueable` execution should be its own transaction and therefore I thought the Logger gets also a new transaction id. I only reconnect them again by setting the `parentLogTransactionId`. And I call `Logger.saveLog()` at several places.

Does anybody have a hint how this could happen or what I might be doing wrong?

<img width="2173" alt="Queueable Logs" src="https://user-images.githubusercontent.com/607607/167399049-ac23e1b8-f5d6-480c-9b25-54812173a389.png">

**Comment by jongpie** [comment]
Hi @reitffunk - that definitely looks like a bug, each execution of the queueable job should have its own `Log__c` record like you described. Do you happen to have any sample code that you could share that would demonstrate this issue? I've tried to reproduce the issue using the queueable class below, but it seems to work as expected every time (each execution of the queueable job creates its own `Log__c` recrd), so I'm not able to reproduce this issue.

```java
public without sharing class Chained_Queueable_Logger_Example implements Queueable {
    private Integer numberOfJobsToChain;
    private String parentLogTransactionId;
    private List<LogEntryEvent__e> logEntryEvents = new List<LogEntryEvent__e>();

    public Chained_Queueable_Logger_Example(Integer numberOfJobsToChain) {
        this(numberOfJobsToChain, null);
    }

    private Chained_Queueable_Logger_Example(Integer numberOfJobsToChain, String parentLogTransactionId) {
        this.numberOfJobsToChain = numberOfJobsToChain;
        this.parentLogTransactionId = parentLogTransactionId;
    }

    public void execute(System.QueueableContext queueableContext) {
        Logger.setParentLogTransactionId(this.parentLogTransactionId);

        Logger.fine('queueableContext==' + queueableContext);
        Logger.info('this.numberOfJobsToChain==' + this.numberOfJobsToChain);
        Logger.info('this.parentLogTransactionId==' + this.parentLogTransactionId);

        Logger.saveLog();

        --this.numberOfJobsToChain;
        if (this.numberOfJobsToChain > 0) {
            String parentLogTransactionId = this.parentLogTransactionId != null ? this.parentLogTransactionId : Logger.getTransactionId();
            System.enqueueJob(new Account_Queueable_Logger_Example(this.numberOfJobsToChain, parentLogTransactionId));
        }
    }
} 
```

**Comment by reitffunk** [comment]
Hi @jongpie,

thanks for testing it directly. Following an observation and some code.

### Different transaction id prefix

First weird thing is that the behavior differs from the Partial-Copy-Sandbox to the Production system. The Logger-Settings are the same and only one "Organization" record exists. Save method ist `EVENT_BUS`.

I just have noticed that the Transaction ID prefix differs for both systems. The "wrong" ones are prefixed with `SLB`. I cannot find anything on Google for `SLB` and currently I also don't have an idea what it means. The correctly separated ones in the partial copy are not prefixed and have it's own transaction ID for every log as expected.

The triggering action comes from an Aura-Component, it is prefixed with `TID` what just means transaction id I guess.

Production-System (unexpected behavior):
<img width="757" alt="image" src="https://user-images.githubusercontent.com/607607/168017015-cb0c51b8-ac55-40fb-9bcb-6da719f8b401.png">

Partial-Copy-Sandbox (expected behavior) (ignore LOG-24, thats another platform event that is logged in this whole process):
<img width="695" alt="image" src="https://user-images.githubusercontent.com/607607/168017319-6b940077-227a-4fda-8544-162bc95c5160.png">

### Code

Now some code, I strip any other logic to focus on triggering the Queueable and the Logger.

Coming from the Aura-Component:
```
public static Map<String, String> updatePositions(Map<String, AM_Bestellposition__c > orderPositions, List<AM_DeliveryPositionDraftApexClass> orderPositionsDeliveries, String deliveryNoteNumber, Date deliveryNoteDate) {
    updateCounterViaQueueable(deliveryPositionDtos, Logger.getTransactionId());
}

private static void updateCounterViaQueueable(List<Delivery_Position_Item__c> deliveryPositionDtos, String loggerTransactionId) {

        UpdateCounterQueueable updateCounterQueueable = new UpdateCounterQueueable(deliveryPositionDtos, 1, loggerTransactionId);
        System.enqueueJob(updateCounterQueueable);
    }
```

The UpdateCounterQueueable() chains itself:

```
        public void execute(QueueableContext ctx) {
            try {
                if (this.parentLogTransactionId != null) {
                    Logger.setParentLogTransactionId(this.parentLogTransactionId);
                }

                // prepare which items to process now and which with the next round of the queueable
                List<Delivery_Position_Item__c> itemsToProcess = new List<Delivery_Position_Item__c>();
                List<Delivery_Position_Item__c> itemsToProcessNextRound = new List<Delivery_Position_Item__c>();

                Integer calculationCounter = this.items.size() > this.batchSize ? this.batchSize : this.items.size();

                for (Integer i = 0; i < calculationCounter; i++) {
                    itemsToProcess.add(this.items.get(i));
                }

                for (Integer i = calculationCounter; i < this.items.size(); i++) {
                    itemsToProcessNextRound.add(this.items.get(i));
                }

                try {
                    this.doTheActualProcessing(itemsToProcess);
                } catch (Exception e) {
                    // We catch it here also. That way the Queueable for the remaining items can
                    // continue.
                    Logger.error('Counter could not be updated: Exception: ' + e.getMessage() +
                            '\nStacktrace: ' + e.getStackTraceString() +
                            '\nItems: ' + JSON.serializePretty(this.items) +
                            '\nItemsToProcess: ' + JSON.serializePretty(itemsToProcess) +
                            '\nItemsToProcessNextRound: ' + JSON.serializePretty(itemsToProcessNextRound)
                    );
                    Logger.saveLog();

                    AM_DeliveryNote_ApexCtrl_Helper.notifyGroupAboutError(itemsToProcess);
                }

                // enqueue the remaining items
                if (itemsToProcessNextRound.size() > 0) {
                    if (itemsToProcessNextRound.size() >= this.batchSize) {
                        System.enqueueJob(new UpdateCounterQueueable(itemsToProcessNextRound, this.batchSize, this.parentLogTransactionId));
                    } else {
                        System.enqueueJob(new UpdateCounterQueueable(itemsToProcessNextRound, itemsToProcessNextRound.size(), this.parentLogTransactionId));
                    }
                }
            }
            catch (Exception ex) {
                Logger.error('Exception: ' + ex.getMessage() + '\nStacktrace: ' + ex.getStackTraceString());
            }
            finally {
                Logger.saveLog();
            }
        }

        private void doTheActualProcessing(List<Delivery_Position_Item__c> objectsToProcess) {
            AM_DeliveryPositionItem_Handler.updateCounters(objectsToProcess);
        }
    }
```


`AM_DeliveryPositionItem_Handler.updateCounters(objectsToProcess);` then calls `Logger.setScenario('Blubb')` (what should be in the execute() I just notice) but it does not make any more calls to `Logger.save()`.

Maybe the surround `try-catch` somehow creates a context between the calls? Should also not be the case, just a desperate shot in the dark :D

**Reply by jongpie** [reply]
@reitffunk this is super helpful, thanks so much for doing this analysis & sharing your findings! Same as you, I'm also not sure what the `SLB` prefix means, but it does indeed seem like the transaction ID is not being reset (which I think is potentially a Salesforce bug, since I'm leveraging the method `System.Request.getCurrent().getRequestId()` that the platform provides). I'm still not 100% sure how to resolve this, but my plan is....

- I'll reach out internally at Salesforce (I'm a Salesforce employee) to see if anyone internally can help explain the behavior that you're seeing with the `SLB`-prefixed transaction IDs in production
- I'm going to consider moving away from using `System.Request.getCurrent().getRequestId()`, and instead go back to using a `UUID` as the transaction ID (before `System.Request.getCurrent().getRequestId()` was introduced in Apex, I previously used a custom `UUID` as the transaction ID) - this is not the first time that the request ID has caused issues, so reverting back to using a custom UUID might be a more stable approach.
  - [I already use a `UUID` in Nebula Logger](https://github.com/jongpie/NebulaLogger/blob/7c6f49c4a7ebed0e24f15725ef2e47d7294ece49/nebula-logger/core/main/logger-engine/classes/Logger.cls#L2735) in a few scenarios where `System.Request.getCurrent().getRequestId()` isn't reliable, so it would be an easy change to use a `UUID` in all scenarios

I don't have an ETA on this yet, but will try to make some progress on this over the next few days. I'll let you know as soon as I have an update!

**Reply by reitffunk** [reply]
Hey @jongpie thank you for being so helpful with your answers and for the motivation to directly implement a change. :)

By the way I was curious from the beginning to why the transaction IDs were often different "formats" for separate `Log__c` records. But from your answer I get it now, it is because they are coming from `System.Request.getCurrent().getRequestId()`. It seemed just simpler to me to always use `UUID`s.

But as I write this I am also curious now, how do you separate the `Log__c` records without looking at the `.getRequestId()`? Or to ask differently, how do you notice that the Logger stays in a new transaction if Salesforce tells you that you are still in the same? I am just wondering because at some point you also must know **when** to create a new `UUID`, right?

**Reply by jongpie** [reply]
Hi @reitffunk, I don't do much directly to handle a new transaction/new request ID - I rely on the platform & Apex's behavior to help with setting a new transaction ID when needed. Specifically, I leverage Apex's way of setting static variables to set some fields only once per transaction - and in Apex, once a transaction ends, any static variables are discarded/they're recreated at the start of a new transaction. I do this for several variables/constants throughout the codebase, but there are 2 relevant variables:

1. [`Logger.TRANSACTION_ID`](https://github.com/jongpie/NebulaLogger/blob/7c6f49c4a7ebed0e24f15725ef2e47d7294ece49/nebula-logger/core/main/logger-engine/classes/Logger.cls#L20) - this is the private constant that's causing issues in your scenario. This is the value returned when you call `Logger.getTransactionId()`. Because it's final, and it's set when instantiated, the value should be the same within a single transaction, and should get a new value when a new transaction starts. However, since you're seeing multiple queueable jobs using the same transaction ID (thus, creating 1 `Log__c`), it makes me think that the platform is re-using values in `System.Request.getCurrent().getRequestId()`
2. [`Logger.currentTransactionEntryNumber`](https://github.com/jongpie/NebulaLogger/blob/7c6f49c4a7ebed0e24f15725ef2e47d7294ece49/nebula-logger/core/main/logger-engine/classes/Logger.cls#L23) - this is a private variable that's used to set the "Entry #" field that you circled in your original screenshot. This variable is similar to `Logger.TRANSACTION_ID` in that it has an initial value (`1`) set when the variable is instantiated - within the same transaction, it is then only incremented; the only time that it ever reset to `1` is when a new transaction starts. So the fact that your screenshot shows multiple `LogEntry__c` records with the same "Entry #" makes me think that this variable is being reset correctly when a new transaction starts, but the request ID is being re-used.

I think in the next release (`v4.7.3`) I'm going to go ahead & change the transaction ID to always be a `UUID` - once that release is ready, would you be able to test it in your production org? My hope is that you will start to see multiple `Log__c` records (1 record per execution of the queueable job).

**Reply by reitffunk** [reply]
Ah, that's interesting regarding the static variables!

And of course I will be able to test it. I am eagerly waiting for the new version - no pressure intended :D

**Reply by jongpie** [reply]
@reitffunk I've just released [v4.7.3](https://github.com/jongpie/NebulaLogger/releases/tag/v4.7.3), which includes the change to only use a `UUID` for the transaction ID. Whenever you have a chance to install it in your prod org, let me know if it resolves your issue with the queueable jobs.

**Reply by reitffunk** [reply]
Hey @jongpie,

I have just looked into the Org and it's working as expected now.

Thank you very much! :)

**Reply by jongpie** [reply]
That's great to hear @reitffunk! Thanks for confirming, and let me know if you run into any other issues

**Comment by jongpie** [comment, marked as answer]
Just noting that this was fixed in [v4.7.3](https://github.com/jongpie/NebulaLogger/releases/tag/v4.7.3) by switching to always using a `UUID` for the transaction ID.
